spec/ruby/optional/capi/hash_spec.rb

```diff
@@ -194,6 +194,44 @@
     end
   end
 
+  describe "rb_hash_bulk_insert" do
+    it 'inserts key-value pairs into the hash' do
+      arr = [:a, 1, :b, 2, :c, 3]
+      hash = {}
+
+      @s.rb_hash_bulk_insert(hash, arr.length, arr)
+
+      hash.should == {a: 1, b: 2, c: 3}
+    end
+
+    it 'overwrites existing keys' do
+      arr = [:a, 4, :b, 5, :c, 6]
+      hash = {a: 1, b: 2}
+
+      @s.rb_hash_bulk_insert(hash, arr.length, arr)
+
+      hash.should == {a: 4, b: 5, c: 6}
+    end
+
+    it 'does not include any keys after the given length' do
+      arr = [:a, 1, :b, 2, :c, 3, :d, 4]
+      hash = {}
+
+      @s.rb_hash_bulk_insert(hash, arr.length - 2, arr)
+
+      hash.should == {a: 1, b: 2, c: 3}
+    end
+
+    it 'does not modify the hash if the length is zero' do
+      arr = []
+      hash = {a: 1, b: 2}
+
+      @s.rb_hash_bulk_insert(hash, arr.length, arr)
+
+      hash.should == {a: 1, b: 2}
+    end
+  end
+
```

Comment:
I would also add the following cases:
- length isn't even - there should be error raised (I suppose)
- how duplicated keys (provided in the array) are handled - I suppose they should just override each other
- it's allowed to pass NULL as array if length argument = 0

Reply:
I'll add tests for the other cases.

As for when the length isn't even, it appears to be that when `RUBY_DEBUG` is not defined MRI accepts an array with an odd number of elements and will read one past the end. Should we allow the same to maintain compatibility?

Reply:
Ah, right. Yes, it makes sense to do the same and to allow passing odd length.

Reply:
We shouldn't add a spec for that odd case, because specs should also pass with `RUBY_DEBUG`, notably `ruby-dev-builder` depends on that.
@andrykonchin Could you add a commit removing that spec?

We could file an issue at https://bugs.ruby-lang.org/, I think it would make sense, it doesn't seem OK the C API reads past the length it's given.